### DISCUSSION TITLE
CPU consumption

### DISCUSSION THREAD

**Discussion by sshvaiko** [directus/directus#23567] [Feature Requests]
### Summary

Hello everyone! We have a Directus on AWS on t3.2xlarge instance (8 vCPUs/32GB RAM). The project is relatively small and doesn't really have any traffic. It uses Redis (AWS ElastiCache) and Postgress (AWS RDS).

It runs on Docker and I will attach a config.

The problem is that we have a lot of issue, when we start making even simple API calls to read from 1 collection without any queries, I see load on CPU in PM2 and then after some requests Directus freezes and starts return 502 error.

We tried to use different configurations, we use latest version but nothing helped. Does anyone have any ideas what can be the case?

We had different Directus instance in Docker (DB + Directus) and it worked well with old version (1 year ago or so).

### Basic Example

```version: "3"
services:
  directus:
    restart: always
    image: directus/directus:11.0.2
    ports:
      - 80:80
    environment:
      CORS_ENABLED: 'true'
      PORT: '80'
      
      CORS_ORIGIN: 'true'
      CORS_METHODS: 'GET,POST,PATCH,DELETE,OPTIONS'
      KEY: '***'
      SECRET: '***'

      DB_CLIENT: 'pg'
      DB_HOST: '***.us-east-1.rds.amazonaws.com'
      DB_PORT: '5432'
      DB_DATABASE: 'postgres'
      DB_USER: '***'
      DB_PASSWORD: '***'

      STORAGE_LOCATIONS: 's3'
      STORAGE_S3_DRIVER: 's3'
      STORAGE_S3_KEY: '***'
      STORAGE_S3_SECRET: '***'
      STORAGE_S3_BUCKET: '***'
      STORAGE_S3_REGION: 'us-east-1'
      STORAGE_S3__ENDPOINT: 's3.us-east-1.amazonaws.com'

      EMAIL_FROM: 'no-reply@***'
      EMAIL_TRANSPORT: 'mailgun'
      EMAIL_MAILGUN_API_KEY: '***'
      EMAIL_MAILGUN_DOMAIN: '***'

      CACHE_ENABLED: 'true'
      CACHE_STORE: 'redis'
      CACHE_AUTO_PURGE: 'true'
      REDIS: 'redis://***.use1.cache.amazonaws.com:6379'

      CONTENT_SECURITY_POLICY_DIRECTIVES__FRAME_SRC: "***"
      CONTENT_SECURITY_POLICY_DIRECTIVES__IMG_SRC: "array:'self',data:,https://*"

      PM2_INSTANCES: '8'
      PM2_AUTO_RESTART: 'true'
      
      PRESSURE_LIMITER_ENABLED: 'false'```

### Motivation

n/a

### Detailed Design

n/a

### Requirements List

n/a

### Drawbacks

n/a

### Alternatives

n/a

### Adoption Strategy

n/a

### Unresolved Questions

_No response_

**Comment by domdus** [comment]
Hard to tell... I see the CPU loads here and there depending on the requests.
Might go into wrong direction but, assuming the AWS/Docker setup is okay:

**1. Cache**
According to your config you set `CACHE_AUTO_PURGE: 'true'`
As much as I like this option (used over years), the more data changing events happen within your directus instance - the cache may become non-existent.
This setting means, whenever a CRUD operation happens ANYWHERE - the cache gets fully (globally, not selective) purged ([see discussion](https://github.com/directus/directus/discussions/12008))

**Example:**
If you track visits of your frequently visited website into a `<visits>` collection - you basically work in a cold start condition always, as the cache is cleared almost every second.

You can improve it a lot with the `CACHE_AUTO_PURGE_IGNORE_LIST` option, unless you don't want to disable `CACHE_AUTO_PURGE`.
Maybe try disabling `CACHE_AUTO_PURGE` in first place to see it fixes the load issue.

----------------------------------------------------------------------

**2. The request**

- How does the API request look like? Literally just `/items/<collection>` ?
- Do you pass request parameters like `limit=-1` to get a ton of items?
- Or any additional env var set like `QUERY_LIMIT_DEFAULT=10000`?
- How many items are in that collection? Does it have many relational fields?
- What is the size of a requests response?
- What happens if you request only certain fields in the query e.g. `fields=id`?

What I want to point out is that performance massively improves when you specify/limit the fields and amount you need in your request if not done yet.

**Example:**
I have a collection `<articles>` with 4000 items, many fields filled with text, tags, translations, categories relations, related content etc..
When I request e.g. `/items/articles?limit=-1` and no fields selected then:

- I see huge load on the CPU (>100%) for long periods
- It takes 13 seconds for the response of the first single uncached request
- The size of the reponse is 27 MB (!) for a single request

If I spam the API with these sorts of requests, especially if they are unique and uncached - then requests will pile up and response times grow to above 40 seconds and minutes until a timeout/cap from nginx eventually.

Once a request is cached and hit again by further requests, the reponse times lower from initially 13 sec (40+ sec pile up) down to 1-2 sec with a lot less load on CPU.

**Reply by sshvaiko** [reply]
Hey! Thanks a lot. I didn't know about the cache behaviour, and we have exactly this case when we write Article Views. I added this table to CACHE_AUTO_PURGE_IGNORE_LIST. Thanks, we will try it!

**Reply by sshvaiko** [reply]
Is there any system tables that you would recommend to add CACHE_AUTO_PURGE_IGNORE_LIST? I see that there are directus_activity,directus_presets in [example](https://docs.directus.io/self-hosted/config-options.html#cache).

**Reply by sshvaiko** [reply]
One one more question :) We currently use t3.medium 4 instances (auto-scale), do you think bigger instances would help?

**Reply by domdus** [reply]
> Is there any system tables that you would recommend to add CACHE_AUTO_PURGE_IGNORE_LIST? I see that there are directus_activity,directus_presets in [example](https://docs.directus.io/self-hosted/config-options.html#cache).

I kept directus_activity,directus_presets in and added basically every collection related to reporting/analytics or like payments & orders - internal stuff.

Anything related to content items & structure (categories, sections) that should update a website/app on change is not included.

**Reply by domdus** [reply]
> One one more question :) We currently use t3.medium 4 instances (auto-scale), do you think bigger instances would help?

Well, it totally depends on the load scenario but 4 of these instances seem quite powerful for API endpoints - as long as the cache is working and the queries are not too heavy. (10000 items at once)

The CPU load issue could still be related to something else in the setup, but I hope you see some improvements with the details from above.
